### ISSUE TITLE
Code review

### ISSUE THREAD

**Issue by kgryte** [compute-io/cartesian-product#1]
Initial implementation for computing a Cartesian product. Only supports `array` inputs.

**Comment by Planeshifter** [comment]
LGTM. One caveat: The error messages do not conform the standard of having as a prefix the function name, parentheses and two colons. For example,

``` javascript
throw new TypeError( 'invalid input argument. All input arguments must be arrays. Value: `' + sets[ i ] + '`.' );
```

should become 

``` javascript
throw new TypeError( 'cartesian()::invalid input argument. All input arguments must be arrays. Value: `' + sets[ i ] + '`.' );
```

In case this was intentional, could you elaborate on the reasoning?

**Comment by kgryte** [comment]
Yes. This was intentional. My thoughts on requiring the function name in the messages have changed. This centers more around maintainability and often begs the question what "name" to use: the module name? the function name? what if the function is just a validation function for a main entry point? and how is the difference between the reported name and the file name seen in stacktraces reconciled? Further, any time a function name is changed, then all references to the function name must be changed, which is not great for maintainability.

Basically, superficially, it seemed like a good idea, but in practice I always look at the file name in the stacktrace along with the line number. Hence, I have effectively deprecated this practice when I make modules. 

A few other practices have changed (nothing requiring major changes), which we can review over conversation.

**Comment by kgryte** [comment]
Also, as long as the function has a name, the function name should appear in the stacktrace anyway.

**Comment by kgryte** [comment]
With this in mind, let me know if you have any further thoughts.

**Comment by Planeshifter** [comment]
No, I think initial version can be published. A few things to consider:
1. Support for Set and Typed Array data types
2. Can you have a look at the `expand.grid` function in R? It allows the creation of a data frame with all combinations of the supplied vectors. Down the road, we might consider something similar.

**Comment by kgryte** [comment]
Yeah, `Set` was already in `TODO.md`, but I had forgotten about `typed arrays`. Thanks!

And will take a look at `expand.grid`.

**Comment by kgryte** [comment]
With [`Set`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set), not particularly enamored with having to deal with iterators. Open question: should the returned `Sets` be of type `Set`, or should they be `arrays`? What if both `Sets` and `arrays` are provided? What is the output data type?

**Comment by kgryte** [comment]
Re: [`expand.grid`](https://stat.ethz.ch/R-manual/R-devel/library/base/html/expand.grid.html). Seems reasonable. Will be dependent on publishing `df` and allowing an output data type option, which is related to the above discussion regarding `sets`, `arrays`, and `typed arrays`.

Along a similar vein, one could argue that a `matrix` could also be provided, where either `rows` or `cols` could be interpreted as `sets`. In which case, the output data structure could be a `matrix`.
